### PR TITLE
Collaboration Whitelisting

### PR DESCRIPTION
**Pull request by carycheng77**

### CODE REVIEW

**Comment by boxcla** [comment]
Verified that @carycheng77 has signed the CLA. Thanks for the pull request!

File: Box.V2.Test.Integration/BoxCollaborationWhitelistManagerTestIntegration.cs [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+    public class BoxCollaborationWhitelistManagerTestIntegration : BoxResourceManagerTestIntegration
+    {
+        [TestMethod]
+        public async Task UserWhitelist_LiveSession_ValidResponse()
+        {
+            var userWhitelist = await _client.CollaborationWhitelistManager.AddCollaborationWhitelistExemptUserAsync("275343214");
+            var retrievedUserWhitelist = await _client.CollaborationWhitelistManager.GetCollaborationWhitelistExemptUserAsync(userWhitelist.Id);
+            var deletedUserWhitelist = await _client.CollaborationWhitelistManager.DeleteCollaborationWhitelistExemptUserAsync(userWhitelist.Id);
+
+            Assert.IsNotNull(userWhitelist, "Was not able to whitelist user");
```
Comment (coolboy) [inline Box.V2.Test.Integration/BoxCollaborationWhitelistManagerTestIntegration.cs]: Assert the content inside, NotNull is not enough

File: Box.V2.Test.Integration/BoxCollaborationWhitelistManagerTestIntegration.cs [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+        [TestMethod]
+        public async Task GetAllUserWhitelist_LiveSession_ValidResponse()
+        {
+            var allUserWhitelistItems = await _client.CollaborationWhitelistManager.GetCollaborationWhitelistExemptUsersAsync();
+            List<BoxCollaborationWhitelistTargetEntry> allUserCollabWhitelist = allUserWhitelistItems.Entries;
+
+            foreach (BoxCollaborationWhitelistTargetEntry userWhitelist in allUserCollabWhitelist)
+            {
+                Assert.IsNotNull(userWhitelist, "collab whitelist was not retrieved");
```
Comment (coolboy) [inline Box.V2.Test.Integration/BoxCollaborationWhitelistManagerTestIntegration.cs]: ditto

File: Box.V2/Managers/BoxCollaborationWhitelistManager.cs [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+            : base(config, service, converter, auth, asUser, suppressNotifications) { }
+
+        /// <summary>
+        /// Used to whitelist a domain for a Box collaborator. You can specify the domain and direction of the whitelist. When whitelisted successfully, only users from the whitelisted
+        /// domain can be invited as a collaborator. 
+        /// </summary>
+        /// <param name="domainToWhitelist">This is the domain to whitelist collaboration.</param>
+        /// <param name="directionForWhitelist">Can be set to inbound, outbound, or both for the direction of the whitelist.</param>
+        /// <returns>The whitelist entry if successfully created.</returns>
+        public async Task<BoxCollaborationWhitelistEntry> AddCollaborationWhitelistEntryAsync(String domainToWhitelist, String directionForWhitelist)
```
Comment (coolboy) [inline Box.V2/Managers/BoxCollaborationWhitelistManager.cs]: If the direction are limited, why not in enum?

File: Box.V2.Test.Integration/BoxCollaborationWhitelistManagerTestIntegration.cs [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+            Assert.IsNotNull(whiteList, "Was not able to create collab whitelist for domain");
+            Assert.IsNotNull(retrievedWhiteList, "Was not able to retrieve collab whitelist by ID");
+            Assert.IsTrue(deletedWhiteList, "Collab whitelist was unable to be deleted");    
+        }
+
+        [TestMethod]
+        public async Task GetAllDomainWhitelist_LiveSession_ValidResponse()
+        {
+            var allCollabWhitelistItems = await _client.CollaborationWhitelistManager.GetAllCollaborationWhitelistEntriesAsync();
```
Comment (coolboy) [inline Box.V2.Test.Integration/BoxCollaborationWhitelistManagerTestIntegration.cs]: We have something called auto paging we need to support and test that

File: Box.V2/Managers/BoxCollaborationWhitelistManager.cs [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+            return response.ResponseObject;
+        }
+
+        /// <summary>
+        /// Used to get information about all domain collaboration whitelists.
+        /// </summary>
+        /// <param name="marker">Position to return results from.</param>
+        /// <param name="limit">Maximum number of entries to return. Default is 100.</param>
+        /// <returns>The collection of domain collaboration whitelist objects is returned.</returns>
+        public async Task<BoxCollaborationWhitelistEntryCollection<BoxCollaborationWhitelistEntry>> GetAllCollaborationWhitelistEntriesAsync(string marker = null, int limit = 100)
```
Comment (coolboy) [inline Box.V2/Managers/BoxCollaborationWhitelistManager.cs]: ditto, autopaging

File: Box.V2/Managers/BoxCollaborationWhitelistManager.cs [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+            return response.ResponseObject;
+        }
+
+        /// <summary>
+        /// Used to get information about all collaboration whitelists for users.
+        /// </summary>
+        /// <param name="marker">Position to return results from.</param>
+        /// <param name="limit">Maximum number of entries to return. Default is 100.</param>
+        /// <returns>The collection of collaboration whitelist object is returned for users.</returns>
+        public async Task<BoxCollaborationWhitelistTargetEntryCollection<BoxCollaborationWhitelistTargetEntry>> GetCollaborationWhitelistExemptUsersAsync(string marker = null, int limit = 100)
```
Comment (coolboy) [inline Box.V2/Managers/BoxCollaborationWhitelistManager.cs]: autopaging and test

File: Box.V2/Models/BoxCollaborationWhitelistEntry.cs [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+        /// The domain to be whitelisted for collaboration.
+        /// </summary>
+        [JsonProperty(PropertyName = FieldDomain)]
+        public String Domain { get; set; }
+
+        /// <summary>
+        /// The direction of the whitelist for collaboration.
+        /// </summary>
+        [JsonProperty(PropertyName = FieldDirection)]
+        public String Direction { get; set; }
```
Comment (coolboy) [inline Box.V2/Models/BoxCollaborationWhitelistEntry.cs]: Direction as Enum?

File: Box.V2/Models/BoxCollaborationWhitelistEntry.cs [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+
+        /// <summary>
+        /// The enterprise the collaboration whitelist belongs to.
+        /// </summary>
+        [JsonProperty(PropertyName = FieldEnterprise)]
+        public BoxEnterprise Enterprise { get; set; }
```
Comment (coolboy) [inline Box.V2/Models/BoxCollaborationWhitelistEntry.cs]: Do we need the complete BoxEnterprise here? or just BoxEntity should be enough?

File: Box.V2/Models/BoxCollaborationWhitelistTargetEntry.cs [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+        public const string FieldUser = "user";
+        public const string FieldEnterprise = "enterprise";
+        public const string FieldCreatedAt = "created_at";
+        public const string FieldModifiedAt = "modified_at";
+
+        /// <summary>
+        /// The user that is exempt from whitelist.
+        /// </summary>
+        [JsonProperty(PropertyName = FieldUser)]
+        public BoxUser User { get; set; }
```
Comment (coolboy) [inline Box.V2/Models/BoxCollaborationWhitelistTargetEntry.cs]: Ditto

File: Box.V2/Models/BoxCollaborationWhitelistTargetEntry.cs [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+
+        /// <summary>
+        /// The domain the whitelist is active in.
+        /// </summary>
+        [JsonProperty(PropertyName = FieldEnterprise)]
+        public BoxEnterprise Enterprise { get; set; }
```
Comment (coolboy) [inline Box.V2/Models/BoxCollaborationWhitelistTargetEntry.cs]: Ditto

**Pull request merged**